provider/CONTRIBUTE.md

```diff
@@ -135,6 +136,120 @@ Once done, you can just press F5 and the debug will start! You can place breakpo
+
+Finally, examples are located in the [examples](https://github.com/terraform-providers/terraform-provider-azurerm/tree/master/examples) folder.
+
+AzureRM Terraform provider uses the [Azure SDK for Go](https://github.com/Azure/azure-sdk-for-go) to interact with Azure. It's important that you respect that rule. If you are trying to do something that is not available in the Azure SDK for Go, then you should open an issue on that repository first.
+
+We recommand that you start with fixing issues/patching an existing data source or resource to understand how it works in details before trying to implement a brand new one.
+
+Don't forget that you are working on your own fork of the repository and that you need to open a pull request to the main repository to bring your update to the HashiCorp team, for review.
+
+### Working with Azure SDK for Go
+
+All the imports for the Azure Go SDK services have to be done in the [config.go](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/config.go) file. It exposes a top level struct, `ArmClient` that exposes all the client that you may use in a data source or resource to interact with Azure.
+
+For example, if you need to work on implementing Azure Batch Account support to the provider:
+
+1. Add import for `"github.com/Azure/azure-sdk-for-go/services/batch/mgmt/2018-12-01/batch"` in the imports list
+2. Add a field for the `batchAccountClient`:
+
+```go
+batchAccountClient     batch.AccountClient
+```
+
+3. Add a function that registers the Azure Batch client:
+
+```go
+func (c *ArmClient) registerBatchClients(endpoint, subscriptionId string, auth autorest.Authorizer) {
+	batchAccount := batch.NewAccountClientWithBaseURI(endpoint, subscriptionId)
+	c.configureClient(&batchAccount.Client, auth)
+	c.batchAccountClient = batchAccount
+}
+```
+
+4. Call the function that registers the Batch client:
+
+```go
+client.registerBatchClients(endpoint, c.SubscriptionID, auth)
+```
+
+By doing the four steps above, you will make sure that when developping the data source or the resource, you can access the Azure SDK for Go clients that you need.
+
+### Developping a Data Source
+
+The file for a data source is named following the convention: data_source_*NAME_OF_THE_DATA_SOURCE*.go.
+
+A data source is composed of two relevant functions:
+
+- The function that creates a `schema.Resource` object that returns the schema (definition) of the data source,i.e. the property that compose the data source, and some rules about those properties (for example, if it is mandatory or not, computed etc.). This function is named by convention dataSourceArm*NAME_OF_THE_DATA_SOURCE*, for example `dataSourceArmBatchAccount` for a Batch Account. 

```

Comment:
Data source schema is much simpler than resource. You only need to provide required properties to locate that resource (for example: [API Management data source](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/data_source_api_management.go#L18-L20)), and mark all other properties returned from the service as `Computed`. There are no `ForceNew`, no `MinItems`, no `MaxItems`, etc.

Reply:
Done.